### ISSUE TITLE
Removing Kindle DRM on a Mac fails

### ISSUE THREAD

**Issue by spirosph** [apprenticeharper/DeDRM_tools#1078]
Hi all.

Using Calibre 4.14.0 and DeDRM_tools_6.7.0 on macOS Catalina, with Kindle for Mac 1.26.1 (also tested with Kindle 1.28.0 with the same results).

I've purchased a Kindle ebook, downloaded on Kindle for Mac (no Kindle e-Ink devices are involved). I followed the instructions at [https://www.mobileread.com/forums/showpost.php?p=3819708&postcount=508](https://www.mobileread.com/forums/showpost.php?p=3819708&postcount=508), and have successfully downloaded the Kindle ebook as an .azw file, imported it to Calibre, but stripping the DRM fails with the error: 

`Found 0 keys to try after 0.3 seconds
Crypto Type is: 2
DeDRM v6.7.0: Failed to decrypt with error: No key found in 0 keys tried.
DeDRM v6.7.0: Looking for new default Kindle Key after 0.3 seconds
No k4Mac kindle-info/rainier/kinf2011 files have been found.`

According to https://www.mobileread.com/forums/showpost.php?p=3819708&postcount=508, this should work. Any ideas/pointers would be greatly appreciated.
Thanks in advance.

[tmpbBIFsa.txt](https://github.com/apprenticeharper/DeDRM_tools/files/4550997/tmpbBIFsa.txt)

**Comment by ElleKayEm** [comment]
Did you have a working setup with Kindle for Mac 1.23 and DeDRM before updating to Catalina?

**Comment by spirosph** [comment]
> Did you have a working setup with Kindle for Mac 1.23 and DeDRM before updating to Catalina?

Thanks for the reply. 
No, I did not have a Kindle for Mac 1.23 + DeDRM setup before. This is the first time I tried DeDRM and I already had Kindle for Mac 1.28 installed (reverted to Kindle for Mac 1.26.1, but same results).

Unfortunately I'm on macOS Catalina, so running 32-bit Kindle for Mac 1.23 is not an option. Any other possible way of doing it?

**Comment by ElleKayEm** [comment]
Not yet.  The tools can't currently decrypt the key for Kindle for Mac for the versions newer than 1.23.  So the method for decrypting books from later versions relies on retrieving a previous key.

**Comment by walkingjew** [comment]
alright - I have a previous setup working on another mac running catalina.
I'm trying to import the old keys to the new macbook. 
but it seems they are not working

is it possible to import old dedrm kindle keys for the same amazon account to another computer that uses the same account? or are they tied up to a particular device?

**Comment by j-howell** [comment]
I don’t think that will work. Kindle for Mac encryption is tied to the serial number of a piece of hardware in your system.

**Comment by walkingjew** [comment]
yeah as I thought - so we are stuck until a new drm tool shows up?

**Comment by spirosph** [comment]
Not having a working Kindle for Mac 1.23/DeDRM setup before updating to Catalina, my solution was to use DeDRM on a "clean" (i.e. Kindle not previously installed) Windows PC I had. 
For those without another "clean" computer to use, a workaround is to set up a Windows 10 VM and use DeDRM there (you can download Windows 10 for free from Microsoft). A hassle I know, but it does the job.

**Comment by walkingjew** [comment]
what about virtualizing a pre-catalina mac os on catalina using vmware fusion? that should do the trick to create the correct hardware bound key for kindle, right?

**Comment by spirosph** [comment]
> what about virtualizing a pre-catalina mac os on catalina using vmware fusion? that should do the trick to create the correct hardware bound key for kindle, right?

Yes, I believe that would do the trick just fine.

**Comment by ElleKayEm** [comment]
Keep in mind that for Kindle for Mac/PC, each new *installation* has its own unique key.  Even for the same machine/user and same Amazon account.  In other words, if you uninstall, reinstall, and need to log in again, you'll have a different key.

(For the E-ink Kindles, keys are tied to the device's serial number.)

**Comment by walkingjew** [comment]
so it won't work? (virtualizing mojave under catalina)

**Comment by ElleKayEm** [comment]
Maybe it would work.  You just need to do it all in the virtual machine...Kindle for Mac and calibre with DeDRM.  Just saying for the apps, the key is not tied to hardware.  It's tied to that specific installation of Kindle for Mac/PC.  If you try it, let us know if it works!

**Comment by j-howell** [comment]
> Just saying for the apps, the key is not tied to hardware. It's tied to that specific installation of Kindle for Mac/PC.

That is true. The key changes with each registration on those platforms.

But the key is itself protected by encryption that depends on the hardware. That is why you cannot just move a .kinf file from one system to another.

**Comment by ElleKayEm** [comment]
Ahh, thanks for the clarification.

**Comment by NickFabry** [comment]
I am running Parallels on Catalina 10.15.5, and I spooled up a Mojave 10.14.6 VM from scratch, specifically for the purpose of running Kindle.app v1.23.1 and DeDRM 6.7.0 on calibre v4.16 - and it works. Of course, you have to sign into the Kindle.app on your VM for this to work.

**Comment by walkingjew** [comment]
yes it does. but you cannot export the encryption keys to catalina. I've tried today, it doesn't work

**Comment by ElleKayEm** [comment]
Thanks for reporting back that it does work as expected.
